python/pylibraft/pylibraft/common/handle.pyx

```diff
@@ -236,3 +236,71 @@ def auto_sync_handle(f):
         handle_docstring=_HANDLE_PARAM_DOCSTRING
     )
     return wrapper
+
+
+cdef class DeviceResourcesSNMG:
+    """
+    DeviceResourcesSNMG manages multi-GPU resources
+    in a single-node setup using RAFT's device_resources_snmg. Refer to
+    the header file raft/core/device_resources_snmg.hpp for interface level
+    details of this struct
+    Parameters
+    ----------
+    device_ids : Optional list to specify which devices will be used
+    Examples
+    --------
+    Basic usage:
+    >>> from pylibraft.common import DeviceResourcesSNMG
+    >>>
+    >>> # to use GPU IDs 0,1,2,3 on machine
+    >>> handle = DeviceResourcesSNMG([0,1,2,3])
+    >>>
+    >>> # to use all GPUs on machine
+    >>> handle = DeviceResourcesSNMG()
+    """
+
+    def __cinit__(self, device_ids=None):
+        self.device_ids = device_ids
+
+        cdef device_resources_snmg* snmg_ptr
+        cdef vector[int] ids
+
+        if self.device_ids is None:
+            snmg_ptr = new device_resources_snmg()
+        else:
+            for id in self.device_ids:
+                ids.push_back(id)
+            snmg_ptr = new device_resources_snmg(ids)
+
+        self.c_obj.reset(snmg_ptr)
+
+    def sync(self):
+        """
+        Issues a sync on the stream set for this instance.
+        """
+        self.c_obj.get()[0].sync_stream()
```

Comment:
What does this mean in the SNMG context?

Reply:
I tried to keep it consistent with the `DeviceResources`. This will sync stream on the main GPU (gpu for `main_gpu_id_` in the cpp layer).

Reply:
Let's remove it then, for now. For me, the expected functionality would be that it would sync all streams for all devices.

Reply:
I think we want to follow the same semantics as all the other `raft::resources`. There's a notion of "main_stream" and "worker_stream". We want to make sure we differentiate between them.

Reply:
From my understanding, it seems like the main GPU will have two streams. The main stream will be synced by this function, however no work will be scheduled on that stream.

The stream on which real work is scheduled for the main GPU will be part of the world resources.

Reply:
@divyegala you are right, and I believe for now the users of this class won't ever really have to do the sync at the python level.

Reply:
Okay cool. So actually in UMAP's case everything but the KNN graph construction will actually run on the main stream, so we should keep this.